src/algebra/algebra/kleene.lean

```diff
@@ -0,0 +1,415 @@
+/-
+Copyright (c) 2022 Siddhartha Prasad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddhartha Prasad
+-/
+import data.real.basic
+import data.vector
+import tactic.explode
+import tactic.find
+import tactic.induction
+import tactic.rcases
+import tactic.rewrite
+
+
+/-!
+# Kleene Algebras
+
+This file defines Kleene algebras, which are used extensively in theory of computation.
+
+A Kleene algebra is an idempotent semiring with an additional unary operator
+called a Kleene star (⋆).
+
+## References
+
+* (Kozen, D. . A completeness theorem for Kleene algebras and the algebra of regular events.)
+  [https://www.cs.cornell.edu/~kozen/Papers/ka.pdf].
+* https://planetmath.org/idempotentsemiring
+* https://encyclopediaofmath.org/wiki/Idempotent_semi-ring
+* https://planetmath.org/kleenealgebra
+
+
+## Tags
+
+kleene algebra
+
+-/
+
+
+
+universe u
+variables {α : Type u}
+
+/--
+An isemiring is a semiring with the additional property that the addition (+)
+operation is idempotent.
+
+For some idempotent semiring S,  ∀ x ∈ S, `x + x = x`.
+
+Additionally, the binary relation ≤ defines a partial order on isemirings.
+where, a ≤ b ↔ a + b = b
+
+-/
+class isemiring  (α : Type u) extends semiring α, semilattice_sup α :=
+(idem_add : ∀ a : α, a + a = a)
+(le_def : ∀ a b : α, a ≤ b ↔ a + b = b)
```

Comment:
My suggestion before (which I was actually waiting for feedback from @YaelDillies for) was to use:
```suggestion
(add_eq_sup : ∀ a : α, a + a = a ⊔ a)
```
which implies many of your results trivially.

Reply:
Yes, fully agree.